### PR TITLE
Convert SelectOptions to TypeScript

### PR DESCRIPTION
**Pull request by nanaya**
The generics don't seem quite right but it at least compiles 🫥

- [x] depends on #9793

### CODE REVIEW

File: resources/js/components/select-options.tsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  private renderOption({ children, onClick, option, selected = false }: ComponentOptionRenderProps<T>) {
+    const cssClasses = classWithModifiers(`${bn}__option`, { selected });
+
+    if (this.props.renderOption != null) {
+      return this.props.renderOption({ children, cssClasses, onClick, option });
+    }
+
+    return (
+      <a
+        key={typeof option.id === 'string' ? option.id : ''}
```
Comment (notbakaneko) [inline resources/js/components/select-options.tsx:128]: can restrict T to `string | number` and not need the `typeof option.id` (and apparently `null` is a valid key but not `undefined` 🤔 )

File: resources/js/components/select-options.tsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  static readonly defaultProps = { blackout: true };
+
+  private readonly blackoutAutoToggleDisposer;
+  private readonly ref = React.createRef<HTMLDivElement>();
+  @observable private showingSelector = false;
+
+  constructor(props: Props<T>) {
+    super(props);
+    makeObservable(this);
+    this.blackoutAutoToggleDisposer = reaction(
```
Comment (notbakaneko) [inline resources/js/components/select-options.tsx:51]: can use `disposeOnUnmount` but not sure if actually better 🤔

File: resources/js/components/select-options.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  private renderOption({ children, onClick, option, selected = false }: ComponentOptionRenderProps<T>) {
+    const cssClasses = classWithModifiers(`${bn}__option`, { selected });
+
+    if (this.props.renderOption != null) {
+      return this.props.renderOption({ children, cssClasses, onClick, option });
+    }
+
+    return (
+      <a
```
Comment (notbakaneko) [inline resources/js/components/select-options.tsx:127]: I can't remember the reason this wasn't changed to `button` 👀

**Review by notbakaneko — CHANGES_REQUESTED** [review]

File: resources/js/components/select-options.tsx (hunk 3 above)
Comment (nanaya) [reply resources/js/components/select-options.tsx:127]: rendering difference aside, button can't contain div 🤔 (can probably also updated but that'll end up with quite a few unrelated updates)

**Review by notbakaneko — CHANGES_REQUESTED** [review]
conflicts 💥

**Review by notbakaneko — APPROVED** [review]

**Pull request merged**